### ISSUE TITLE
evolve with mutil-gpu

### ISSUE THREAD

**Issue by alicera** [ultralytics/yolov5#6599]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

bash run.sh
```bash
#!/bin/bash
# Multi-GPU
for i in 0 1 2 3; do
  nohup python train.py --epochs 300 --data data/4cls.yaml --weights d11b_large-object-bright_20220207_best.pt --evolve --device $i > evolve_gpu_$i.log &
done
```
![image](https://user-images.githubusercontent.com/59388989/153386467-e43bec4b-360f-45de-9ee4-e25cb287f551.png)
![image](https://user-images.githubusercontent.com/59388989/153386500-ba29a602-da26-45d8-ba34-caab8709c8ef.png)

**Comment by glenn-jocher** [comment]
@alicera what's the question?

**Comment by alicera** [comment]
Does it success to run for see the message?

**Comment by alicera** [comment]
Command
```
python train.py --epochs 300 --data coco128.yaml --weights yolov5s.pt --cache --evolve
```
![image](https://user-images.githubusercontent.com/59388989/153792356-aa70159e-cfcc-4853-8c2d-bc824070c901.png)

every generation as runs/evolve/hyp_evolved.yam 
When is the `hyp_evolved.yam` generated? Until 300 epoch ?

**Comment by glenn-jocher** [comment]
@alicera hyp_evolved.yaml is updated every generation with the latest results.

**Comment by alicera** [comment]
I already run 70 epoch and the folder `runs/evolve/exp/` only has the `results.csv`.
So my training have some problem?

**Comment by glenn-jocher** [comment]
@alicera that's not correct. Evolution is very easy, you can start just by attaching the `--evolve` argument to the Colab notebook train cell. As you can see there will be 3 files in your runs/evolve/exp directory which are updated every generation:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb

<img width="2158" alt="Screenshot 2022-02-15 at 10 49 26" src="https://user-images.githubusercontent.com/26833433/154037085-68b19c3b-1458-465f-8494-0b385b4b6124.png">

**Comment by alicera** [comment]
`you can see there will be 3 files in your runs/evolve/exp directory which are updated every generation:`
The result is used [colab]( https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) to generate?

If I use the command `python train.py --epochs 300 --data coco128.yaml --weights yolov5s.pt --evolve
` on the linux terminal, it is also work ? 
Because I use it on the terminal, it still no generate evolve.csv and hyp_evolve.yaml. But there is result.csv.

**Comment by glenn-jocher** [comment]
@alicera yes example is shown in Colab. Evolution works in every environment, just like any normal YOLOv5 training.

**Comment by zaipinai** [comment]
> @alicera yes example is shown in Colab. Evolution works in every environment, just like any normal YOLOv5 training.

I have a same problem with alicera. After training 150 epochs(I set it max to 150), it created hyp.yaml and hyp_evolved.yaml. It seems that the two are reversed, the hyp.yaml is more likely to be the result.
1.hyp.yaml 2.hyp_evolved.yaml 3.my original hyp
![image](https://user-images.githubusercontent.com/35369552/157679361-d3a0fb8f-4fa8-4643-bb28-4ba3bc527186.png)
![image](https://user-images.githubusercontent.com/35369552/157679434-8466f1b7-c728-4bb0-9107-d3b2e6b48e0c.png)
![image](https://user-images.githubusercontent.com/35369552/157679480-f8f8bfa2-8e21-4b18-937e-1f6d68c2ca56.png)

**Comment by glenn-jocher** [comment]
@zaipinai good news 😃! Your original issue may now be fixed ✅ in a previous PR #6604. Current code no longer outputs hyp.yaml and applies an indexing fix that caused incorrect final results to display:

<img width="1932" alt="Screenshot 2022-03-10 at 15 43 24" src="https://user-images.githubusercontent.com/26833433/157685763-21350e25-406d-4b65-9f26-4ee1a3bcdfb0.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by zaipinai** [comment]
> @zaipinai good news 😃! Your original issue may now be fixed ✅ in a previous PR #6604. Current code no longer outputs hyp.yaml and applies an indexing fix that caused incorrect final results to display:
> 
> <img alt="Screenshot 2022-03-10 at 15 43 24" width="1932" src="https://user-images.githubusercontent.com/26833433/157685763-21350e25-406d-4b65-9f26-4ee1a3bcdfb0.png">
> 
> To receive this update:
> 
> * **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
> * **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
> * **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> * **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image [![Docker Pulls](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

Thanks for your help!  I use modify YOLO v5 with transformer prediction head and it doesn't update. I will update it later. Sorry for bothering you for this mistake.

**Comment by pderrenger** [comment]
@zaipinai no problem at all! Feel free to update your modified version and let us know if you encounter any issues. We're here to help with any YOLOv5-related questions you might have.
